src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -675,21 +685,43 @@ private class FastqDemultiplexer(val sampleInfos: Seq[SampleInfo],
     // Get the sample
     val (sampleInfo, numMismatches) = matchSampleBarcode(subReads)
 
-    // Create the [[DemuxRecord]]s.
-    def opt(s: String): Option[String] = if (s.isEmpty) None else Some(s)
-    val molecularBarcode = subReads.filter(_.kind == SegmentType.MolecularBarcode).map(_.bases).mkString("-")
-    val sampleBarcode    = subReads.filter(_.kind == SegmentType.SampleBarcode).map(_.bases).mkString("-")
-    val readName         = reads.head.name
-    var readNumber       = 0
+    // Get all the comments
     val comments = reads.flatMap(_.comment)
-    val records          = subReads.filter(_.kind == SegmentType.Template).map { read =>
-      val comment = if (comments.isEmpty) None else Some(comments(readNumber))
-      readNumber += 1
-      DemuxRecord(readName, read.bases, read.quals, opt(molecularBarcode), opt(sampleBarcode), readNumber, pairedEnd, comment)
+
+    // Method to get all the bases of a given type
+    def bases(segmentType: SegmentType): Option[String] = {
+      val b = subReads.filter(_.kind == segmentType).map(_.bases).mkString("-")
+      if (b.isEmpty) None else Some(b)
     }
 
+    // Get the molecular and sample barcodes
+    val molecularBarcode = bases(SegmentType.MolecularBarcode)
+    val sampleBarcode    = bases(SegmentType.SampleBarcode)
+
+    val demuxRecords = reads.zip(this.variableReadStructures)
+      .map { case (read, rs) => rs.extract(read.bases, read.quals) }
+      .filter { segments => segments.exists(_.kind == SegmentType.Template) }
+      .zipWithIndex
+      .map { case (segments, readIndex) =>
+        val readNumber = readIndex + 1
+        val comment    = if (comments.isEmpty) None else Some(comments(readIndex))
+        val read       = segments.find(_.kind == SegmentType.Template).getOrElse(unreachable("Bug: should contain a template segment"))
```

Comment:
Do you validate elsewhere that each read can contain only 0-1 template segments?  If not, maybe here it should validate that you get exactly one?